compiler/lib/flow_lexer.ml

```diff
@@ -450,6 +456,62 @@ let new_line env lexbuf =
   let lex_bol = { Lex_env.line = Lex_env.line env + 1; offset } in
   { env with Lex_env.lex_bol }
 
+let decode_identifier =
+  let sub_lexeme lexbuf trim_start trim_end =
+    Sedlexing.Utf8.sub_lexeme
+      lexbuf
+      trim_start
+      (Sedlexing.lexeme_length lexbuf - trim_start - trim_end)
+  in
+  let rec id_char buf lexbuf =
+    match%sedlex lexbuf with
+    | unicode_escape ->
+        let hex = sub_lexeme lexbuf 2 0 in
+        let code = int_of_string ("0x" ^ hex) in
+        if not (Uchar.is_valid code) then failwith "invalid unicode";
+        Buffer.add_utf_8_uchar buf (Uchar.of_int code);
+        id_char buf lexbuf
+    | codepoint_escape ->
+        let hex = sub_lexeme lexbuf 3 1 in
+        let code = int_of_string ("0x" ^ hex) in
+        if not (Uchar.is_valid code) then failwith "invalid unicode";
```

Comment:
You cannot really do that, if you get a high surrogate you need to combine it with the next low surrogate to make the final uchar character ([`UTF16SurrogatePairToCodePoint`](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#sec-utf16decodesurrogatepair) in ECMA). See for example the combination logic [here](https://github.com/ocaml/ocaml/blob/fb4e2acbd4e61fd376d3ef6bdf886ad6055b190f/stdlib/bytes.ml#L744-L759).

For example '🐫' can be represented by the escape sequence '\xD83D\xDC2B',

Reply:
I'm not sure non-scalar escape sequence are accepted in identifier. node doesn't accept `\u{d83b}\u{de62}` for instance.

Reply:
I suspect that are not in `\u{}` (even though nothing seems to prevent them in ECMA) but are in `\uXXXX` did you try ?

Reply:
note that I don't think 🐫(`U+1F42B`) is a valid identifier. '𞹢' (`U+1ee62`) seems to be and is accepted by node.

Reply:
Ah but that's the reason: 
```
> ucharinfo 🐫 -k id_continue
ID_Continue: false
```
However:

```
> ucharinfo U+1D4DD -k name -k id_continue -k UTF-16BE
Name: MATHEMATICAL BOLD SCRIPT CAPITAL N
ID_Continue: true
UTF-16BE: \xD8\x35\xDC\xDD
```

`var nat_is_\uD835\uDCDD = true` doesn't work but `var nat_is_\u{1D4DD} = true` does (e.g. in in the chrome console). But I can't reconcile that with my reading of: 

> In other words, if a \ [UnicodeEscapeSequence](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#prod-UnicodeEscapeSequence) sequence were replaced by the [SourceCharacter](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#prod-SourceCharacter) it contributes, the result must still be a valid [IdentifierName](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#prod-IdentifierName) that has the exact same sequence of [SourceCharacter](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#prod-SourceCharacter) elements as the original [IdentifierName](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#prod-IdentifierName).

[here](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#sec-identifier-names).

Reply:
Ah I think I get the convoluted logic :-)

The definition of [`SourceCharacter`](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#prod-SourceCharacter) is: 


> [SourceCharacter](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#prod-SourceCharacter) ::
any Unicode code point

(and not a *scalar value*). So going back to my previous quote: an escape of a surrogate contributes a surrogate code point `SourceCharacter`. But the `ID_Continue` property of surrogates is (I suppose, `ucharinfo` can only query scalar values) `false`, so it cannot be part of an identifier. So surrogate pairs escape sequences are excluded from the identifier syntax.

Reply:
it seems that surrogate pairs are not combined. So, `\u{12345}` is a valid identifier, but `\ud808\udf45` is not.

Reply:
I've updated the lexer to better handle escaped surrogate pairs (record errors and recover) and added expect tests.